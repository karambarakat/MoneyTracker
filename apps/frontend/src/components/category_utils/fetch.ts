import { UseQueryResult, useQuery } from '@tanstack/react-query'
import { useMemo } from 'react'

export const iconQuery = async () => {
  const res = await import('./icons')

  return res.default.split('\n') as string[]
}

/**
 *
 * @param whichOne select specific icon or return all icons
 */
export function useIcon(whichOne: string) {
  const res = useQuery({
    queryFn: iconQuery,
    suspense: false,
    queryKey: ['icons'],
    cacheTime: 1000 * 60 * 60 * 24,
  }) as UseQueryResult<string>
  // return data of type string please

  const validIndex = useMemo(() => {
    const target = Number(whichOne)

    if (isNaN(target) || target > 876 || target < 0)
      return new Error('Invalid index')

    return target
  }, [whichOne])

  if (validIndex instanceof Error) return validIndex

  if (res.status === 'success')
    return { status: 'success', data: res.data[validIndex] }

  return res
}

export const defaultValue =
  'm38.49,32h-14.49V8h24v18h-6.04l-3.46,6Zm-18.49,4v-14c-7.73,0-14,6.27-14,14s6.27,14,14,14,14-6.27,14-14h-14Zm27.73-6h-3.46l-12.12,21c.77,1.33.96,1.67,1.73,3h24.25c.77-1.33.96-1.67,1.73-3l-12.12-21Z'
