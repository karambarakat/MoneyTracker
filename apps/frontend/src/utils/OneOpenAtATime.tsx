import { useId } from '@mantine/hooks'
import { createContext, useContext, useEffect, useState } from 'react'
import { WithChildren } from 'ui/src/utils/WithChildren'

const context = createContext<{
  open: false | string
  setOpen: (s: string | false) => void
}>({ open: false, setOpen: () => console.error('no context assigned') })

export const useOneState = (initial?: boolean) => {
  const id = useId()
  const { open, setOpen } = useContext(context)
  const [_open, _setOpen] = useState(open === id)
  useEffect(() => {
    if (initial) setOpen(id)
  }, [initial])
  useEffect(() => {
    if (open === id) _setOpen(true)
    else _setOpen(false)
  }, [open])

  return [
    _open,
    function (arg: boolean | ((old: boolean) => boolean)) {
      if (typeof arg === 'function') {
        const rt = arg(_open)
        return rt ? setOpen(id) : setOpen(false)
      }
      if (arg) setOpen(id)
      else setOpen(false)
    },
  ] as const
}

export const OneStateProvider = ({ children }: WithChildren) => {
  const [open, setOpen] = useState<false | string>(false)
  return (
    <context.Provider value={{ open, setOpen }}>{children}</context.Provider>
  )
}
