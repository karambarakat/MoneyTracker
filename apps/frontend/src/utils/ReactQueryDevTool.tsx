import { ReactQueryDevtools } from '@tanstack/react-query-devtools'
import React from 'react'

const ReactQueryDevtoolsProductionLazy = React.lazy(() =>
  import('@tanstack/react-query-devtools/build/lib/index.prod.js').then(d => ({
    default: d.ReactQueryDevtools,
  })),
)

export function ReactQueryDevtoolsProduction() {
  const [showDevtools, setShowDevtools] = React.useState(false)

  React.useEffect(() => {
    // @ts-ignore
    window.toggleDevtools = () => setShowDevtools(old => !old)
  }, [])

  return showDevtools ? (
    <React.Suspense fallback={null}>
      <ReactQueryDevtoolsProductionLazy />
    </React.Suspense>
  ) : null
}

export default function ReactQueryDevTools_() {
  return (
    <>
      <ReactQueryDevtools initialIsOpen={false} position="bottom-right" />
      <ReactQueryDevtoolsProduction />
    </>
  )
}
