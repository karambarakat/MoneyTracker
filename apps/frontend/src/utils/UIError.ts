export default class UIError extends Error {
  jsError: unknown
  constructor(message: string, jsError?: unknown) {
    super(message)
    this.jsError = jsError
  }
}
