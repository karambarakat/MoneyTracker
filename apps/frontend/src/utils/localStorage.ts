import { useEffect, useState } from 'react'
import { Jwt_ } from 'types/backend'
import { User } from 'types/gql/graphql'

const eventTarget = new EventTarget()

class LocalItem<T> {
  /**
   * @private
   */
  _item: T | undefined
  key: string
  constructor(key: string) {
    this.key = key
    const localItem = localStorage.getItem(key)
    this._item = localItem ? JSON.parse(localItem as string) : undefined
  }

  setItem(item: T | undefined) {
    if (!item) localStorage.removeItem(this.key)
    else localStorage.setItem(this.key, JSON.stringify(item))

    this._item = item

    // React
    eventTarget.dispatchEvent(new Event(this.key))
  }

  getItem() {
    const item = localStorage.getItem(this.key)

    if (!item) return

    return JSON.parse(item) as T
  }

  use() {
    const [item, setItem] = useState<LocalItem<T>>(this)

    useEffect(() => {
      const handler = () => setItem(new LocalItem(this.key))
      eventTarget.addEventListener(this.key, handler)
      return () => eventTarget.removeEventListener(this.key, handler)
    }, [])

    return item
  }
}

export const profile = new LocalItem<User>('profile')

export const token = Object.assign(new LocalItem<string>('token'), {
  valid(this: LocalItem<string>) {
    const item = this.getItem()

    if (!item) return false

    return (
      Number((JSON.parse(atob(item.split('.')[1])) as Jwt_).exp) >=
      Date.now() / 1000
    )
  },
})
