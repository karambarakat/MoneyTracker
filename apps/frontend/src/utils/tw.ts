// this is temprorary until I switch to UnoCSS or
// figure out a way to simplify tw configuration completely

import tw from 'twin.macro'

export const iconRed = tw`fill-red-800 dark:fill-red-400`

export const iconBlue = tw`fill-blue-800 dark:fill-blue-400`
