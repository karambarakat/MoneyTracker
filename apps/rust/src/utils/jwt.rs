#[allow(dead_code)]
// for some limitation for ts_rs, this exists for the sake of simplicity
#[derive(ts_rs::TS)]
#[ts(export)]
pub struct Jwt_ {
    pub id: String,
    pub email: String,
    pub user_id: String,
    pub exp: String,
    pub iat: String,
}

pub struct Jwt {
    pub id: String,
    pub email: String,
    pub user_id: String,
    pub exp: DateTime<Utc>,
    pub iat: DateTime<Utc>,
}
use chrono::DateTime;
use chrono::Duration;
use chrono::Utc;
use hmac::{Hmac, Mac};
use jwt::SignWithKey;
use jwt::VerifyWithKey;
use sha2::Sha256;
use std::collections::BTreeMap;

use crate::errors::MyErrors;

impl Jwt {
    pub fn validate(str: &str) -> Result<Self, MyErrors> {
        let env = std::env::var("JWT_SALT").expect("JWT_SALT is not set");

        let key: Hmac<Sha256> = Hmac::new_from_slice(env.as_bytes())
            .map_err(|err| MyErrors::InternalError(Box::new(err)))?;

        let claims: BTreeMap<String, String> = str
            .verify_with_key(&key)
            .map_err(|err| MyErrors::InternalError(Box::new(err)))?;

        let exp = claims.get("exp");

        let exp = match exp {
            Some(exp) => {
                // test if expirted
                let exp = exp
                    .parse::<i64>()
                    .map_err(|err| MyErrors::InternalError(Box::new(err)))?;
                let exp = DateTime::<Utc>::from_utc(
                    chrono::NaiveDateTime::from_timestamp_opt(exp, 0).unwrap(),
                    Utc,
                );

                if exp < Utc::now() {
                    return Err(MyErrors::ExpiredBearerToken);
                }

                exp
            }
            None => return Err(MyErrors::ExpiredBearerToken),
        };

        Ok(Self {
            id: claims.get("id").unwrap().to_string(),
            email: claims.get("email").unwrap().to_string(),
            user_id: claims.get("user_id").unwrap().to_string(),
            exp,
            iat: DateTime::<Utc>::from_utc(
                chrono::NaiveDateTime::from_timestamp_opt(
                    claims
                        .get("iat")
                        .unwrap()
                        .parse::<i64>()
                        .map_err(|err| MyErrors::InternalError(Box::new(err)))?,
                    0,
                )
                .unwrap(),
                Utc,
            ),
        })
    }

    pub fn sign(id: &str, email: &str, user_id: &str) -> Result<String, MyErrors> {
        let env = std::env::var("JWT_SALT").expect("JWT_SALT is not set");

        let key: Hmac<Sha256> = Hmac::new_from_slice(env.as_bytes())
            .map_err(|err| MyErrors::InternalError(Box::new(err)))?;

        let claims = {
            let mut claims: BTreeMap<String, String> = BTreeMap::new();

            let now = Utc::now();
            let exp = now + Duration::days(1);

            claims.insert("id".to_string(), id.to_string());
            claims.insert("email".to_string(), email.to_string());
            claims.insert("user_id".to_string(), user_id.to_string());
            claims.insert("exp".to_string(), exp.timestamp().to_string());
            claims.insert("iat".to_string(), now.timestamp().to_string());

            claims
        };

        let token_str = claims
            .sign_with_key(&key)
            .map_err(|err| MyErrors::InternalError(Box::new(err)))?;

        Ok(token_str)
    }
}

impl std::fmt::Display for Jwt {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "[User: {},****]", self.email)
    }
}
