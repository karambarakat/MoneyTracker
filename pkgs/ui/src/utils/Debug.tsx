import tw from 'twin.macro'
import { WithChildren } from './WithChildren'

export default function Debug(p: WithChildren) {
  return (
    <pre
      className="debug"
      css={{
        '& > *': tw`mb-1`,
        '& > :last-child': tw`mb-0`,
        button: {
          ...tw`bg-teal-500/50 px-2`,
          '&:hover': tw`bg-teal-500/70`,
        },
      }}
      tw="p-3 absolute top-10 right-10 bg-teal-500/20 w-[min-content]"
    >
      {p.children}
    </pre>
  )
}
