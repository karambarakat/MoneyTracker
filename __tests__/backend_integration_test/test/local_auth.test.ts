import { it, expect, describe, beforeAll, afterAll } from '@jest/globals'
import fetch from 'node-fetch'
import { HELPER_SERVER, REST_API } from '../utils/constants'
import { expectResponse, profileRest } from '../utils/expect'

const ctx: any = {
  ids: [],
}

describe('basic token', () => {
  it('protected by token', async () => {
    const input = [REST_API + '/auth/local', {}] as const
    const res = await fetch(...input).then(res => res.json())

    expect(res).toMatchSnapshot()
  })

  it('protected by token: login', async () => {
    const input = [
      REST_API + '/auth/local/login',
      {
        method: 'POST',
      },
    ] satisfies Parameters<typeof fetch>
    const res = await fetch(...input).then(res => res.json())

    expect(res).toMatchSnapshot()
  })

  it('protected by token: register', async () => {
    const input = [
      REST_API + '/auth/local/register',
      {
        method: 'POST',
      },
    ] satisfies Parameters<typeof fetch>
    const res = await fetch(...input).then(res => res.json())

    expect(res).toMatchSnapshot()
  })

  it('bad basic token', async () => {
    const input = [
      REST_API + '/auth/local/register',
      {
        method: 'POST',
        headers: {
          authorization: 'Basic badToken',
        },
      },
    ] satisfies Parameters<typeof fetch>

    const res = await fetch(...input).then(res => res.json())

    expect(res).toMatchSnapshot()
  })

  it('not a basic token', async () => {
    const input = [
      REST_API + '/auth/local/register',
      {
        method: 'POST',
        headers: {
          authorization: 'Bearer badToken',
        },
      },
    ] satisfies Parameters<typeof fetch>

    const res = await fetch(...input).then(res => res.json())

    expect(res).toMatchSnapshot()
  })

  it('good basic token', async () => {
    const input = [
      REST_API + '/auth/local/register',
      {
        method: 'POST',
        headers: {
          content: 'application/json',
          authorization:
            'Basic ' + Buffer.from('admin:pass').toString('base64'),
        },
        body: JSON.stringify({}),
      },
    ] satisfies Parameters<typeof fetch>

    const res = await fetch(...input)
    const json: any = await res.json()

    expectResponse(res)
    ctx.user = Object.assign({}, json.data)
    ctx.ids.push(json.data.id)
    ctx.user.token = res.headers.get('x-token')

    expect(json.data.email).toBe('admin')

    expect(json.data).toMatchSnapshot(profileRest)
  })
  console // left intentionally
})

describe('local auth', () => {
  it('login', async () => {
    const input = [
      REST_API + '/auth/local/login',
      {
        method: 'POST',
        headers: {
          content: 'application/json',
          authorization:
            'Basic ' + Buffer.from('admin:pass').toString('base64'),
        },
        body: JSON.stringify({}),
      },
    ] satisfies Parameters<typeof fetch>

    const res = await fetch(...input)
    const json: any = await res.json()

    expectResponse(res)

    expect(json.data).toMatchSnapshot(profileRest)
  })

  it('usable returned token', async () => {
    const input = [
      REST_API + '/graphql',
      {
        method: 'POST',
        body: JSON.stringify({
          query: 'query {}',
          variables: {},
        }),
        headers: {
          content: 'application/json',
          authorization: `Bearer ${ctx.user.token}`,
        },
      },
    ] satisfies Parameters<typeof fetch>

    const res = await fetch(...input)

    expectResponse(res)
  })

  console // left intentionally
})

afterAll(async () => {
  await Promise.all(
    ctx.ids.map((id: string) =>
      fetch(HELPER_SERVER + '/remove_user_data/' + id),
    ),
  )
})
