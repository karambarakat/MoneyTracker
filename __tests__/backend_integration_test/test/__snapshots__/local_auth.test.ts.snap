// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic token bad basic token 1`] = `
Object {
  "data": null,
  "error": Object {
    "code": "BadRequest",
    "info": Object {
      "reason": "token decoding failed, input is valid?",
    },
    "message": "Ops, some error occured, please try again",
    "status": 400,
  },
}
`;

exports[`basic token good basic token 1`] = `
Object {
  "avatar": null,
  "created_at": Any<Number>,
  "display_name": null,
  "email": "admin",
  "id": Any<String>,
  "providers": "local",
  "updated_at": Any<Number>,
}
`;

exports[`basic token not a basic token 1`] = `
Object {
  "data": null,
  "error": Object {
    "code": "BadRequest",
    "info": Object {
      "reason": "not a basic token",
    },
    "message": "Ops, some error occured, please try again",
    "status": 400,
  },
}
`;

exports[`basic token protected by token 1`] = `
Object {
  "data": null,
  "error": Object {
    "code": "BadRequest",
    "info": Object {
      "reason": "header is not provided",
    },
    "message": "Ops, some error occured, please try again",
    "status": 400,
  },
}
`;

exports[`basic token protected by token: login 1`] = `
Object {
  "data": null,
  "error": Object {
    "code": "BadRequest",
    "info": Object {
      "reason": "header is not provided",
    },
    "message": "Ops, some error occured, please try again",
    "status": 400,
  },
}
`;

exports[`basic token protected by token: register 1`] = `
Object {
  "data": null,
  "error": Object {
    "code": "BadRequest",
    "info": Object {
      "reason": "header is not provided",
    },
    "message": "Ops, some error occured, please try again",
    "status": 400,
  },
}
`;

exports[`local auth login 1`] = `
Object {
  "avatar": null,
  "created_at": Any<Number>,
  "display_name": null,
  "email": "admin",
  "id": Any<String>,
  "providers": "local",
  "updated_at": Any<Number>,
}
`;
